README.md
# ComS_327_Loading-Pokemon

COM S 327, Spring 2022
Programming Project 1.07
Loading Pokemon ´
We’ve parsed in a number of data files that specify certain details about how to create pokemon. Now ´
we’re actually going to load them into our game. We’ll add the ability to encounter them, but we’ll save
battling and capturing for next week.
We’re going to simplify many of the mechanics of the pokemon main series games (MSGs) in our im- ´
plementation. The full mechanics are just too tedious, which would result in a lot of code and not much
learning. If correct (or more correct) mechanics are important to you, you’re welcome to implement something more detailed than presented here.
When walking through tall grass, each move the PC makes has a certain chance of encountering a
pokemon. In the MSGs, different areas of the world will have different kinds of pok ´ emon in the grass, as ´
well as other types of terrain (desert, caves, water) that may spawn pokemon, as well. We’ll simplify this ´
by allowing any (literally any!) pokemon to spawn in any tall grass, this includes fully evolved forms and ´
legendary and mythical pokemon. We’ve parsed in details on 1092 pok ´ emon; we’ll select from them on a ´
uniform distribution, which means that you have an equal chance of encountering a pidgey and a mewtwo.
As for the probability of encountering a pokemon in the tall grass, something like 10% seems reasonable. ´
The levels of encountered pokemon will increase with the manhatten distance from the origin. When ´
distance is less than or equal to 200, minimum level is 1 and maximum level is distance / 2 (integer division).
When distance exceeds 200, minimum level becomes (distance - 200) / 2 and maximum level is 100.
Encountered pokemon will know up to two moves. The known moves are restricted to the pok ´ emon’s ´
level-up learnset and will only be less than two if the pokemon has fewer than two moves in its level-up ´
learnset at its level.
The pokemon in our database are grouped in many ways. The ´ version group id in pokemon
moves.csv is a unique identifier of a pokemon game, and (I think) a value of 19 in the only one for ´
which all pokemon have learnsets. We’ll use that one, so that we can use all of the pok ´ emon. Again, you’re ´
welcome to do something different here, but it will increase the complication of your game somewhat.
To find the level-up moveset of a pokemon, you’ll want ´ pokemon moves.version group id
equal to 19, pokemon moves.pokemon id equal to pokemon.species id and pokemon
moves.pokemon move method id equal to 1. From these, select your two moves from a uniform
distribution.
Pokemon stats (attack, defense, etc.) are given in ´ pokemon stats.csv, which we didn’t parse last
week, so you’ll need to parse that in now. The mapping of stat numbers to names is given in stats.csv.
Each pokemon has an additive deviation from its species’ base stats, known as IVs (individual values). IVs ´
for each stat are uniformly distributed in the range [0,15]. Generate IVs for each spawned pokemon for each ´
of HP, attack, defense, speed, special attack, and special defense.
The mechanics of pokemon gender are somewhat involved (designed to reduce storage, see ´
https://bulbapedia.bulbagarden.net/wiki/Individual values). Some pokemon don’t even have genders. We’ll ´
simplify this by making all species have equal probability of being male or female. As with everything else
here, you’re welcome to implement the more complex mechanic if you like.
Pokemon have a 1/8192 chance of being shiny, based on their IVs (see the above link). Again, simplify ´
this by ignoring the IVs and simply using, e.g., rand() % 8192 == 0. Feel free to reduce that modulus to
increase your shiny rate.
In order to level up your pokemon, we’ll apply the following formulae: ´
HP = b
((base + IV ) × 2) × level
100
c + level + 10
Otherstat = b
((base + IV ) × 2) × level
100
c + 5
where base is the pokemon species’ base stat, ´ IV is its IV for the given stat and level is the pokemon’s ´
level. Again, we’re simplifying an MSG mechanic here by ignoring EVs (effort values), which normally
add an additional boost to pokemon growth. See ´
https://bulbapedia.bulbagarden.net/wiki/Stat for the deets.
When encountering pokemon, you’ll add a placeholder function for the battle and capture sequence in ´
which you’ll print the pokemon’s level, stats, and moves and pause for input (escape, any key, etc.). ´
A final note: There are probably only two of you (maybe three) who know less about pokemon than me. ´
Please let me know if I got something qualitatively wrong here, or if you think of a better simplification to
MSG mechanics than I’ve proposed.


didic_elmin.assignment-1.07/CHANGELOG
FEB 2 2022
ADDED World Generation in PrintWorld()

FEB 9 2022
ADDED World grid and added the correct movement inputs and allowed users to move around

FEB 23 2022
ADDED HIKERS AND RIVALS AND FIND THE PATH FROM EVERY SPOT ON THE MAP USING A* 
CHANGED ALOT OF CODE TO USE PROFS CODE FROM PDF

MAR 10 2022
ADDED MOVEMENT FOR TRAINERS

MAR 24 2022
ADDED NCURSES AND MOVEMENT
YOU CAN NOT ENTER A POKEMART OR CENTER OR CROSS MAPS

MAR 31 2022
ADDED COMMANDS TO PRINT OUR CSV FILES

APRIL 06 2022
10% CHANCE TO FIND POKEMON IN TALL GRASS

didic_elmin.assignment-1.07/db_parse.h
#ifndef DB_PARSE_H
# define DB_PARSE_H

struct pokemon_db {
  int id;
  char identifier[30];
  int species_id;
  int height;
  int weight;
  int base_experience;
  int order;
  int is_default;
};

struct move_db {
  int id;
  char identifier[30];
  int generation_id;
  int type_id;
  int power;
  int pp;
  int accuracy;
  int priority;
  int target_id;
  int damage_class_id;
  int effect_id;
  int effect_chance;
  int contest_type_id;
  int contest_effect_id;
  int super_contest_effect_id;
};

struct pokemon_move_db {
  int pokemon_id;
  int version_group_id;
  int move_id;
  int pokemon_move_method_id;
  int level;
  int order;
};

struct pokemon_species_db {
  int id;
  char identifier[30];
  int generation_id;
  int evolves_from_species_id;
  int evolution_chain_id;
  int color_id;
  int shape_id;
  int habitat_id;
  int gender_rate;
  int capture_rate;
  int base_happiness;
  int is_baby;
  int hatch_counter;
  int has_gender_differences;
  int growth_rate_id;
  int forms_switchable;
  int is_legendary;
  int is_mythical;
  int order;
  int conquest_order;
};

struct experience_db {
  int growth_rate_id;
  int level;
  int experience;
};

struct pokemon_stats {
  int pokemon_id;
  int stat_id;
  int base_stat;
  int effort;

};

extern pokemon_move_db pokemon_moves[528239];
extern pokemon_db pokemon[1093];
extern char *types[19];
extern move_db moves[845];
extern pokemon_species_db species[899];
extern experience_db experience[601];
extern pokemon_stats stats[10194];

void db_parse(bool print);

#endif


didic_elmin.assignment-1.07/io.cpp
#include <unistd.h>
#include <ncurses.h>
#include <ctype.h>
#include <stdlib.h>
#include <limits.h>
#include <math.h>

#include "io.h"
#include "character.h"
#include "poke327.h"
#include "db_parse.h"

typedef struct io_message {
  /* Will print " --more-- " at end of line when another message follows. *
   * Leave 10 extra spaces for that.                                      */
  char msg[71];
  struct io_message *next;
} io_message_t;




static io_message_t *io_head, *io_tail;

void io_init_terminal(void)
{
  initscr();
  raw();
  noecho();
  curs_set(0);
  keypad(stdscr, TRUE);
  start_color();
  init_pair(COLOR_RED, COLOR_RED, COLOR_BLACK);
  init_pair(COLOR_GREEN, COLOR_GREEN, COLOR_BLACK);
  init_pair(COLOR_YELLOW, COLOR_YELLOW, COLOR_BLACK);
  init_pair(COLOR_BLUE, COLOR_BLUE, COLOR_BLACK);
  init_pair(COLOR_MAGENTA, COLOR_MAGENTA, COLOR_BLACK);
  init_pair(COLOR_CYAN, COLOR_CYAN, COLOR_BLACK);
  init_pair(COLOR_WHITE, COLOR_WHITE, COLOR_BLACK);
}

void io_reset_terminal(void)
{
  endwin();

  while (io_head) {
    io_tail = io_head;
    io_head = io_head->next;
    free(io_tail);
  }
  io_tail = NULL;
}

void io_queue_message(const char *format, ...)
{
  io_message_t *tmp;
  va_list ap;

  if (!(tmp = (io_message_t *) malloc(sizeof (*tmp)))) {
    perror("malloc");
    exit(1);
  }

  tmp->next = NULL;

  va_start(ap, format);

  vsnprintf(tmp->msg, sizeof (tmp->msg), format, ap);

  va_end(ap);

  if (!io_head) {
    io_head = io_tail = tmp;
  } else {
    io_tail->next = tmp;
    io_tail = tmp;
  }
}

static void io_print_message_queue(uint32_t y, uint32_t x)
{
  while (io_head) {
    io_tail = io_head;
    attron(COLOR_PAIR(COLOR_CYAN));
    mvprintw(y, x, "%-80s", io_head->msg);
    attroff(COLOR_PAIR(COLOR_CYAN));
    io_head = io_head->next;
    if (io_head) {
      attron(COLOR_PAIR(COLOR_CYAN));
      mvprintw(y, x + 70, "%10s", " --more-- ");
      attroff(COLOR_PAIR(COLOR_CYAN));
      refresh();
      getch();
    }
    free(io_tail);
  }
  io_tail = NULL;
}

/**************************************************************************
 * Compares trainer distances from the PC according to the rival distance *
 * map.  This gives the approximate distance that the PC must travel to   *
 * get to the trainer (doesn't account for crossing buildings).  This is  *
 * not the distance from the NPC to the PC unless the NPC is a rival.     *
 *                                                                        *
 * Not a bug.                                                             *
 **************************************************************************/
static int compare_trainer_distance(const void *v1, const void *v2)
{
  const Character *const *c1 = (const Character *const *) v1;
  const Character *const *c2 = (const Character *const *) v2;

  return (world.rival_dist[(*c1)->pos[dim_y]][(*c1)->pos[dim_x]] -
          world.rival_dist[(*c2)->pos[dim_y]][(*c2)->pos[dim_x]]);
}

static Character *io_nearest_visible_trainer()
{
  Character **c, *n;
  uint32_t x, y, count;

  c = (Character **) malloc(world.cur_map->num_trainers * sizeof (*c));

  /* Get a linear list of trainers */
  for (count = 0, y = 1; y < MAP_Y - 1; y++) {
    for (x = 1; x < MAP_X - 1; x++) {
      if (world.cur_map->cmap[y][x] && world.cur_map->cmap[y][x] !=
          &world.pc) {
        c[count++] = world.cur_map->cmap[y][x];
      }
    }
  }

  /* Sort it by distance from PC */
  qsort(c, count, sizeof (*c), compare_trainer_distance);

  n = c[0];

  free(c);

  return n;
}

void io_display()
{
  uint32_t y, x;
  Character *c;

  clear();
  for (y = 0; y < MAP_Y; y++) {
    for (x = 0; x < MAP_X; x++) {
      if (world.cur_map->cmap[y][x]) {
        mvaddch(y + 1, x, world.cur_map->cmap[y][x]->symbol);
      } else {
        switch (world.cur_map->map[y][x]) {
        case ter_boulder:
        case ter_mountain:
          attron(COLOR_PAIR(COLOR_MAGENTA));
          mvaddch(y + 1, x, '%');
          attroff(COLOR_PAIR(COLOR_MAGENTA));
          break;
        case ter_tree:
        case ter_forest:
          attron(COLOR_PAIR(COLOR_GREEN));
          mvaddch(y + 1, x, '^');
          attroff(COLOR_PAIR(COLOR_GREEN));
          break;
        case ter_path:
        case ter_exit:
          attron(COLOR_PAIR(COLOR_YELLOW));
          mvaddch(y + 1, x, '#');
          attroff(COLOR_PAIR(COLOR_YELLOW));
          break;
        case ter_mart:
          attron(COLOR_PAIR(COLOR_BLUE));
          mvaddch(y + 1, x, 'M');
          attroff(COLOR_PAIR(COLOR_BLUE));
          break;
        case ter_center:
          attron(COLOR_PAIR(COLOR_RED));
          mvaddch(y + 1, x, 'C');
          attroff(COLOR_PAIR(COLOR_RED));
          break;
        case ter_grass:
          attron(COLOR_PAIR(COLOR_GREEN));
          mvaddch(y + 1, x, ':');
          attroff(COLOR_PAIR(COLOR_GREEN));
          break;
        case ter_clearing:
          attron(COLOR_PAIR(COLOR_GREEN));
          mvaddch(y + 1, x, '.');
          attroff(COLOR_PAIR(COLOR_GREEN));
          break;
        default:
 /* Use zero as an error symbol, since it stands out somewhat, and it's *
  * not otherwise used.                                                 */
          attron(COLOR_PAIR(COLOR_CYAN));
          mvaddch(y + 1, x, '0');
          attroff(COLOR_PAIR(COLOR_CYAN)); 
       }
      }
    }
  }

  mvprintw(23, 1, "PC position is (%2d,%2d) on map %d%cx%d%c.",
           world.pc.pos[dim_x],
           world.pc.pos[dim_y],
           abs(world.cur_idx[dim_x] - (WORLD_SIZE / 2)),
           world.cur_idx[dim_x] - (WORLD_SIZE / 2) >= 0 ? 'E' : 'W',
           abs(world.cur_idx[dim_y] - (WORLD_SIZE / 2)),
           world.cur_idx[dim_y] - (WORLD_SIZE / 2) <= 0 ? 'N' : 'S');
  mvprintw(22, 1, "%d known %s.", world.cur_map->num_trainers,
           world.cur_map->num_trainers > 1 ? "trainers" : "trainer");
  mvprintw(22, 30, "Nearest visible trainer: ");
  if ((c = io_nearest_visible_trainer())) {
    attron(COLOR_PAIR(COLOR_RED));
    mvprintw(22, 55, "%c at %d %c by %d %c.",
             c->symbol,
             abs(c->pos[dim_y] - world.pc.pos[dim_y]),
             ((c->pos[dim_y] - world.pc.pos[dim_y]) <= 0 ?
              'N' : 'S'),
             abs(c->pos[dim_x] - world.pc.pos[dim_x]),
             ((c->pos[dim_x] - world.pc.pos[dim_x]) <= 0 ?
              'W' : 'E'));
    attroff(COLOR_PAIR(COLOR_RED));
  } else {
    attron(COLOR_PAIR(COLOR_BLUE));
    mvprintw(22, 55, "NONE.");
    attroff(COLOR_PAIR(COLOR_BLUE));
  }

  io_print_message_queue(0, 0);

  refresh();
}

uint32_t io_teleport_pc(pair_t dest)
{
  /* Just for fun. And debugging.  Mostly debugging. */

  do {
    dest[dim_x] = rand_range(1, MAP_X - 2);
    dest[dim_y] = rand_range(1, MAP_Y - 2);
  } while (world.cur_map->cmap[dest[dim_y]][dest[dim_x]]                  ||
           move_cost[char_pc][world.cur_map->map[dest[dim_y]]
                                                [dest[dim_x]]] == INT_MAX ||
           world.rival_dist[dest[dim_y]][dest[dim_x]] < 0);

  return 0;
}

static void io_scroll_trainer_list(char (*s)[40], uint32_t count)
{
  uint32_t offset;
  uint32_t i;

  offset = 0;

  while (1) {
    for (i = 0; i < 13; i++) {
      mvprintw(i + 6, 19, " %-40s ", s[i + offset]);
    }
    switch (getch()) {
    case KEY_UP:
      if (offset) {
        offset--;
      }
      break;
    case KEY_DOWN:
      if (offset < (count - 13)) {
        offset++;
      }
      break;
    case 27:
      return;
    }

  }
}

static void io_list_trainers_display(Npc **c,
                                     uint32_t count)
{
  uint32_t i;
  char (*s)[40]; /* pointer to array of 40 char */

  s = (char (*)[40]) malloc(count * sizeof (*s));

  mvprintw(3, 19, " %-40s ", "");
  /* Borrow the first element of our array for this string: */
  snprintf(s[0], 40, "You know of %d trainers:", count);
  mvprintw(4, 19, " %-40s ", s[0]);
  mvprintw(5, 19, " %-40s ", "");

  for (i = 0; i < count; i++) {
    snprintf(s[i], 40, "%16s %c: %2d %s by %2d %s",
             char_type_name[c[i]->ctype],
             c[i]->symbol,
             abs(c[i]->pos[dim_y] - world.pc.pos[dim_y]),
             ((c[i]->pos[dim_y] - world.pc.pos[dim_y]) <= 0 ?
              "North" : "South"),
             abs(c[i]->pos[dim_x] - world.pc.pos[dim_x]),
             ((c[i]->pos[dim_x] - world.pc.pos[dim_x]) <= 0 ?
              "West" : "East"));
    if (count <= 13) {
      /* Handle the non-scrolling case right here. *
       * Scrolling in another function.            */
      mvprintw(i + 6, 19, " %-40s ", s[i]);
    }
  }

  if (count <= 13) {
    mvprintw(count + 6, 19, " %-40s ", "");
    mvprintw(count + 7, 19, " %-40s ", "Hit escape to continue.");
    while (getch() != 27 /* escape */)
      ;
  } else {
    mvprintw(19, 19, " %-40s ", "");
    mvprintw(20, 19, " %-40s ",
             "Arrows to scroll, escape to continue.");
    io_scroll_trainer_list(s, count);
  }

  free(s);
}

static void io_list_trainers()
{
  Character **c;
  uint32_t x, y, count;

  c = (Character **) malloc(world.cur_map->num_trainers * sizeof (*c));

  /* Get a linear list of trainers */
  for (count = 0, y = 1; y < MAP_Y - 1; y++) {
    for (x = 1; x < MAP_X - 1; x++) {
      if (world.cur_map->cmap[y][x] && world.cur_map->cmap[y][x] !=
          &world.pc) {
        c[count++] = world.cur_map->cmap[y][x];
      }
    }
  }

  /* Sort it by distance from PC */
  qsort(c, count, sizeof (*c), compare_trainer_distance);

  /* Display it */
  io_list_trainers_display((Npc **)(c), count);
  free(c);

  /* And redraw the map */
  io_display();
}

void io_pokemart()
{
  mvprintw(0, 0, "Welcome to the Pokemart.  Could I interest you in some Pokeballs?");
  refresh();
  getch();
}

void io_poke_battle()
{
   clear();

  double man_distance = abs(world.cur_idx[dim_x] - 200) + abs(world.cur_idx[dim_y] - 200);

  int IVhp = rand() % 15;
  int IVattack = rand() % 15;
  int IVdefense = rand() % 15;
  int IVspecialattack = rand() % 15;
  int IVspecialdefense = rand() % 15;
  int IVspeed = rand() % 15;




  int level;
  int randa = 1;
  bool shiny = false;
  bool gender = false;
  randa = (rand() % 898);
int start = 0;

    for( start = 0; start < 10194; start++)
    {
      if(stats[start].pokemon_id == randa)
      break;
    }

    if(man_distance == 1 || man_distance == 0) {
        level = 1;
    }
    else if(man_distance > 200) {
      
        level = (rand() % (int)((100 - ((man_distance - 200) / 2) + 1)) + ((man_distance - 200) / 2));
    }
    else {
        level = (rand() % (int)(man_distance / 2)) + 1;
    }

    int basehp = stats[start].base_stat;
    int baseattack = stats[start + 1].base_stat;
    int basedefense = stats[start + 2].base_stat;
    int basespecialattack = stats[start + 3].base_stat;
    int basesepcialdefense = stats[start + 4].base_stat;
    int basespeed = stats[start + 5].base_stat;

    int hp = (((basehp + IVhp) * 2) * level) / 100 + level + 10;
    int attack = (((baseattack + IVattack) * 2 * level) / 100 + 5);
    int defense = (((basedefense + IVdefense) * 2 * level) / 100 + 5);
    int special_attack = (((basespecialattack + IVspecialattack) * 2 * level) / 100 + 5);
    int special_defense = (((basesepcialdefense + IVspecialdefense) * 2 * level) / 100 + 5);
    int speed = (((basespeed + IVspeed) * 2 * level) / 100 + 5);


    

     int moves_start = -1;
     int moves_end = -1;
    
    for(int i = 0; i < 528238; i++) {
        if(pokemon_moves[i].pokemon_id == pokemon[randa].species_id) {
          
            moves_start = i;
      
            break;
        }
    }
    for(int i = moves_start; i < 528238; i++) {
        if(pokemon_moves[i].pokemon_id != pokemon[randa].species_id) {
    
            moves_end = i;
       
            break;
        }
        else if(i == 528238 - 1) {
            moves_end = 528238;
            break;
        }
    }
int move_one_id = 0;
int  move_two_id = 0;

    for(int i = moves_start; i < moves_end; i++) {
        if(pokemon_moves[i].pokemon_move_method_id == 1 && pokemon_moves[i].level <= level && pokemon_moves[i].level != 0) {
             move_one_id = pokemon_moves[i].move_id;
             break;
        }
    }

    for(int i = moves_start; i < moves_end; i++) {
        if(pokemon_moves[i].pokemon_move_method_id == 1 && pokemon_moves[i].level <= level && pokemon_moves[i].level != 0) {
               move_two_id = pokemon_moves[i].move_id + 1;
              
        }
    }
    

    
   
    

 if(rand() % 2 == 0)
 {
   gender = true;
 }

 if(rand() % 8192 == 0)
 {
   shiny = true;
 }

 
  mvprintw(0,0, "A wild %s appeard!", pokemon[randa].identifier);
  mvprintw(1,0, "level: %d", level);
  mvprintw(2,0, "Move 1: %s", moves[move_one_id].identifier);
  mvprintw(3,0, "Move 2: %s", moves[move_two_id].identifier);
  mvprintw(4,0, "id: %d", pokemon[randa].id);
  mvprintw(5,0, "is shiny: %s", shiny ? "true" : "false");
  mvprintw(6,0, "gender: %s", gender ? "female" : "male");
  mvprintw(7,0, "HP: %d  IV %d", hp, IVhp);
  mvprintw(8,0, "Attack: %d  IV %d", attack, IVattack);
  mvprintw(9,0, "Defense: %d IV %d", defense, IVdefense);
  mvprintw(10,0, "Special Attack: %d IV %d", special_attack, IVspecialattack);
  mvprintw(11,0, "Special Defense: %d IV %d", special_defense, IVdefense);
  mvprintw(12,0, "Speed: %d IV %d", speed, IVspeed);

  refresh();
  getch();
}

void io_pokemon_center()
{
  mvprintw(0, 0, "Welcome to the Pokemon Center.  How can Nurse Joy assist you?");
  refresh();
  getch();
}

void io_battle(Character *aggressor, Character *defender)
{
  Npc *npc;

  io_display();
  mvprintw(0, 0, "Aww, how'd you get so strong?  You and your pokemon must share a special bond!");
  refresh();
  getch();
  if (!(npc = dynamic_cast<Npc *>(aggressor))) {
    npc = dynamic_cast<Npc *>(defender);
  }
  
  npc->defeated = 1;
  if (npc->ctype == char_hiker || npc->ctype == char_rival) {
    npc->mtype = move_wander;
  }
}

uint32_t move_pc_dir(uint32_t input, pair_t dest)
{
  dest[dim_y] = world.pc.pos[dim_y];
  dest[dim_x] = world.pc.pos[dim_x];

  switch (input) {
  case 1:
  case 2:
  case 3:
    dest[dim_y]++;
    break;
  case 4:
  case 5:
  case 6:
    break;
  case 7:
  case 8:
  case 9:
    dest[dim_y]--;
    break;
  }
  switch (input) {
  case 1:
  case 4:
  case 7:
    dest[dim_x]--;
    break;
  case 2:
  case 5:
  case 8:
    break;
  case 3:
  case 6:
  case 9:
    dest[dim_x]++;
    break;
  case '>':
    if (world.cur_map->map[world.pc.pos[dim_y]][world.pc.pos[dim_x]] ==
        ter_mart) {
      io_pokemart();
    }
    if (world.cur_map->map[world.pc.pos[dim_y]][world.pc.pos[dim_x]] ==
        ter_center) {
      io_pokemon_center();
    }
    break;
  }

  if ((world.cur_map->map[dest[dim_y]][dest[dim_x]] == ter_exit) &&
      (input == 1 || input == 3 || input == 7 || input == 9)) {
    // Exiting diagonally leads to complicated entry into the new map
    // in order to avoid INT_MAX move costs in the destination.
    // Most easily solved by disallowing such entries here.
    return 1;
  }

 if (world.cur_map->map[world.pc.pos[dim_y]][world.pc.pos[dim_x]] == ter_grass) {
    
    //call func
    if(rand() % 10 == 0)
    {
        io_poke_battle();
    }
      
  }



  if (world.cur_map->cmap[dest[dim_y]][dest[dim_x]]) {
    if (dynamic_cast<Npc *>(world.cur_map->cmap[dest[dim_y]][dest[dim_x]]) &&
        ((Npc *) world.cur_map->cmap[dest[dim_y]][dest[dim_x]])->defeated) {
      // Some kind of greeting here would be nice
      return 1;
    } else if (dynamic_cast<Npc *>
               (world.cur_map->cmap[dest[dim_y]][dest[dim_x]])) {
      io_battle(&world.pc, world.cur_map->cmap[dest[dim_y]][dest[dim_x]]);
      // Not actually moving, so set dest back to PC position
      dest[dim_x] = world.pc.pos[dim_x];
      dest[dim_y] = world.pc.pos[dim_y];
    }
  }
  
  if (move_cost[char_pc][world.cur_map->map[dest[dim_y]][dest[dim_x]]] ==
      INT_MAX) {
    return 1;
  }

  return 0;
}


 void io_teleport_world(pair_t dest)
{
  int x , y;
  
  world.cur_map->cmap[world.pc.pos[dim_y]][world.pc.pos[dim_x]] = NULL;

  mvprintw(0, 0, "Enter x [-200, 200]: ");
  refresh();
  echo();
  curs_set(1);
  mvscanw(0, 21, (char *)"%d", &x);
  mvprintw(0, 0, "Enter y [-200, 200]:          ");
  refresh();
  mvscanw(0, 21,  (char *)"%d", &y);
  refresh();
  noecho();
  curs_set(0);

  if (x < -200) {
    x = -200;
  }
  if (x > 200) {
    x = 200;
  }
  if (y < -200) {
    y = -200;
  }
  if (y > 200) {
    y = 200;
  }
  
  x += 200;
  y += 200;

  world.cur_idx[dim_x] = x;
  world.cur_idx[dim_y] = y;

  new_map(1);
  io_teleport_pc(dest);
}


void io_handle_input(pair_t dest)
{
  uint32_t turn_not_consumed;
  int key;

  do {
    switch (key = getch()) {
    case '7':
    case 'y':
    case KEY_HOME:
      turn_not_consumed = move_pc_dir(7, dest);
      break;
    case '8':
    case 'k':
    case KEY_UP:
      turn_not_consumed = move_pc_dir(8, dest);
      break;
    case '9':
    case 'u':
    case KEY_PPAGE:
      turn_not_consumed = move_pc_dir(9, dest);
      break;
    case '6':
    case 'l':
    case KEY_RIGHT:
      turn_not_consumed = move_pc_dir(6, dest);
      break;
    case '3':
    case 'n':
    case KEY_NPAGE:
      turn_not_consumed = move_pc_dir(3, dest);
      break;
    case '2':
    case 'j':
    case KEY_DOWN:
      turn_not_consumed = move_pc_dir(2, dest);
      break;
    case '1':
    case 'b':
    case KEY_END:
      turn_not_consumed = move_pc_dir(1, dest);
      break;
    case '4':
    case 'h':
    case KEY_LEFT:
      turn_not_consumed = move_pc_dir(4, dest);
      break;
    case '5':
    case ' ':
    case '.':
    case KEY_B2:
      dest[dim_y] = world.pc.pos[dim_y];
      dest[dim_x] = world.pc.pos[dim_x];
      turn_not_consumed = 0;
      break;
    case '>':
      turn_not_consumed = move_pc_dir('>', dest);
      break;
    case 'Q':
      dest[dim_y] = world.pc.pos[dim_y];
      dest[dim_x] = world.pc.pos[dim_x];
      world.quit = 1;
      turn_not_consumed = 0;
      break;
      break;
    case 't':
      /* Teleport the PC to a random place in the map.              */
      io_teleport_pc(dest);
      turn_not_consumed = 0;
      break;
    case 'T':
      /* Teleport the PC to any map in the world.                   */
      io_teleport_world(dest);
      turn_not_consumed = 0;
      break;
    case 'm':
      io_list_trainers();
      turn_not_consumed = 1;
      break;
    case 'q':
      /* Demonstrate use of the message queue.  You can use this for *
       * printf()-style debugging (though gdb is probably a better   *
       * option.  Not that it matters, but using this command will   *
       * waste a turn.  Set turn_not_consumed to 1 and you should be *
       * able to figure out why I did it that way.                   */
      io_queue_message("This is the first message.");
      io_queue_message("Since there are multiple messages, "
                       "you will see \"more\" prompts.");
      io_queue_message("You can use any key to advance through messages.");
      io_queue_message("Normal gameplay will not resume until the queue "
                       "is empty.");
      io_queue_message("Long lines will be truncated, not wrapped.");
      io_queue_message("io_queue_message() is variadic and handles "
                       "all printf() conversion specifiers.");
      io_queue_message("Did you see %s?", "what I did there");
      io_queue_message("When the last message is displayed, there will "
                       "be no \"more\" prompt.");
      io_queue_message("Have fun!  And happy printing!");
      io_queue_message("Oh!  And use 'Q' to quit!");

      dest[dim_y] = world.pc.pos[dim_y];
      dest[dim_x] = world.pc.pos[dim_x];
      turn_not_consumed = 0;
      break;
    default:
      /* Also not in the spec.  It's not always easy to figure out what *
       * key code corresponds with a given keystroke.  Print out any    *
       * unhandled key here.  Not only does it give a visual error      *
       * indicator, but it also gives an integer value that can be used *
       * for that key in this (or other) switch statements.  Printed in *
       * octal, with the leading zero, because ncurses.h lists codes in *
       * octal, thus allowing us to do reverse lookups.  If a key has a *
       * name defined in the header, you can use the name here, else    *
       * you can directly use the octal value.                          */
      mvprintw(0, 0, "Unbound key: %#o ", key);
      turn_not_consumed = 1;
    }
    refresh();
  } while (turn_not_consumed);
}
